### ISSUE TITLE
wrong reprojection to EPSG:3857

### ISSUE THREAD

**Issue by f-christophe** [deegree/deegree3#653]
Hi,

There is a problem when reprojecting to EPSG:3857. We use this get request against a deegree 3.3.17 WFS implementation, the features are stored (SQL feature store) as EPSG:4326:

`http://deegree-server/deegree-webservices-3.3.17/services/bu?SERVICE=WFS&VERSION=2.0.0&request=GetFeature&typeNames=bu-core2d:Building&Filter=<Filter><PropertyIsEqualTo><ValueReference>gml:identifier</ValueReference><Literal>http://inspire.ngi.be/so/BE/NGI/ITGI-Vref/bu/Building/co_building/C2BA8B1E-AD7F-486D-B22E-6381108EEE0C</Literal></PropertyIsEqualTo></Filter>&srsName=EPSG:3857`

We get this response:

```xml
<?xml version='1.0' encoding='UTF-8'?>
<wfs:FeatureCollection xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.opengis.net/wfs/2.0 http://schemas.opengis.net/wfs/2.0/wfs.xsd http://www.opengis.net/gml/3.2 http://schemas.opengis.net/gml/3.2.1/gml.xsd http://inspire.ec.europa.eu/schemas/bu-core2d/3.0 http://ign-sd-wfs-1:8080/deegree-webservices-3.3.17/services/bu?SERVICE=WFS&amp;VERSION=2.0.0&amp;REQUEST=DescribeFeatureType&amp;OUTPUTFORMAT=application%2Fgml%2Bxml%3B+version%3D3.2&amp;TYPENAME=bu-core2d:Building&amp;NAMESPACES=xmlns(bu-core2d,http%3A%2F%2Finspire.ec.europa.eu%2Fschemas%2Fbu-core2d%2F3.0)" xmlns:wfs="http://www.opengis.net/wfs/2.0" timeStamp="2015-12-14T09:09:23Z" xmlns:gml="http://www.opengis.net/gml/3.2" numberMatched="unknown" numberReturned="0">
  <!--NOTE: numberReturned attribute should be 'unknown' as well, but this would not validate against the current version of the WFS 2.0 schema (change upcoming). See change request (CR 144): https://portal.opengeospatial.org/files?artifact_id=43925.-->
  <wfs:member>
    <bu-core2d:Building xmlns:bu-core2d="http://inspire.ec.europa.eu/schemas/bu-core2d/3.0" gml:id="BU-CORE2D_BUILDING_0f14e287-6f2b-42b0-a55e-4e2a760e903d">
      <gml:identifier codeSpace="http://inspire.ec.europa.eu/ids">http://inspire.ngi.be/so/BE/NGI/ITGI-Vref/bu/Building/co_building/C2BA8B1E-AD7F-486D-B22E-6381108EEE0C</gml:identifier>
      <bu-base:beginLifespanVersion xmlns:bu-base="http://inspire.ec.europa.eu/schemas/bu-base/3.0">2012-11-19T00:00:00</bu-base:beginLifespanVersion>
      <bu-base:conditionOfConstruction xmlns:bu-base="http://inspire.ec.europa.eu/schemas/bu-base/3.0"/>
      <bu-base:inspireId xmlns:bu-base="http://inspire.ec.europa.eu/schemas/bu-base/3.0">
        <base:Identifier xmlns:base="http://inspire.ec.europa.eu/schemas/base/3.3">
          <base:localId>bu.Building.co_building.C2BA8B1E-AD7F-486D-B22E-6381108EEE0C</base:localId>
          <base:namespace>BE.NGI.ITGI-Vref</base:namespace>
        </base:Identifier>
      </bu-base:inspireId>
      <bu-core2d:geometry2D>
        <bu-base:BuildingGeometry2D xmlns:bu-base="http://inspire.ec.europa.eu/schemas/bu-base/3.0">
          <bu-base:geometry>
            <gml:Polygon gml:id="BU-CORE2D_BUILDING_0f14e287-6f2b-42b0-a55e-4e2a760e903d_BU-CORE2D_GEOMETRY2D" srsName="EPSG:3857">
              <gml:exterior>
                <gml:LinearRing>
                  <gml:posList>507054.821 6590370.120 507058.563 6590360.196 507051.137 6590357.393 507047.393 6590367.317 507054.821 6590370.120</gml:posList>
                </gml:LinearRing>
              </gml:exterior>
            </gml:Polygon>
          </bu-base:geometry>
          <bu-base:referenceGeometry>true</bu-base:referenceGeometry>
          <bu-base:horizontalGeometryReference xmlns:xlink="http://www.w3.org/1999/xlink" xlink:href="http://inspire.ec.europa.eu/codelist/HorizontalGeometryReferenceValue/roofEdge"/>
          <bu-base:verticalGeometryReference xmlns:xlink="http://www.w3.org/1999/xlink" xlink:href="http://inspire.ec.europa.eu/codelist/ElevationReferenceValue/generalRoofEdge"/>
          <bu-base:horizontalGeometryEstimatedAccuracy uom="m">1.25</bu-base:horizontalGeometryEstimatedAccuracy>
          <bu-base:verticalGeometryEstimatedAccuracy uom="m">0.35</bu-base:verticalGeometryEstimatedAccuracy>
        </bu-base:BuildingGeometry2D>
      </bu-core2d:geometry2D>
    </bu-core2d:Building>
  </wfs:member>
</wfs:FeatureCollection>
```
The coordinates are wrong, they should be the same as when reprojecting to EPSG:900931:

```xml
<?xml version='1.0' encoding='UTF-8'?>
<wfs:FeatureCollection xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.opengis.net/wfs/2.0 http://schemas.opengis.net/wfs/2.0/wfs.xsd http://www.opengis.net/gml/3.2 http://schemas.opengis.net/gml/3.2.1/gml.xsd http://inspire.ec.europa.eu/schemas/bu-core2d/3.0 http://inspire.ngi.be/wfs/bu?SERVICE=WFS&amp;VERSION=2.0.0&amp;REQUEST=DescribeFeatureType&amp;OUTPUTFORMAT=application%2Fgml%2Bxml%3B+version%3D3.2&amp;TYPENAME=bu-core2d:Building&amp;NAMESPACES=xmlns(bu-core2d,http%3A%2F%2Finspire.ec.europa.eu%2Fschemas%2Fbu-core2d%2F3.0)" xmlns:wfs="http://www.opengis.net/wfs/2.0" timeStamp="2015-12-14T08:47:17Z" xmlns:gml="http://www.opengis.net/gml/3.2" numberMatched="unknown" numberReturned="0">
  <!--NOTE: numberReturned attribute should be 'unknown' as well, but this would not validate against the current version of the WFS 2.0 schema (change upcoming). See change request (CR 144): https://portal.opengeospatial.org/files?artifact_id=43925.-->
  <wfs:member>
    <bu-core2d:Building xmlns:bu-core2d="http://inspire.ec.europa.eu/schemas/bu-core2d/3.0" gml:id="BU-CORE2D_BUILDING_0f14e287-6f2b-42b0-a55e-4e2a760e903d">
      <gml:identifier codeSpace="http://inspire.ec.europa.eu/ids">http://inspire.ngi.be/so/BE/NGI/ITGI-Vref/bu/Building/co_building/C2BA8B1E-AD7F-486D-B22E-6381108EEE0C</gml:identifier>
      <bu-base:beginLifespanVersion xmlns:bu-base="http://inspire.ec.europa.eu/schemas/bu-base/3.0">2012-11-19T00:00:00</bu-base:beginLifespanVersion>
      <bu-base:conditionOfConstruction xmlns:bu-base="http://inspire.ec.europa.eu/schemas/bu-base/3.0"/>
      <bu-base:inspireId xmlns:bu-base="http://inspire.ec.europa.eu/schemas/bu-base/3.0">
        <base:Identifier xmlns:base="http://inspire.ec.europa.eu/schemas/base/3.3">
          <base:localId>bu.Building.co_building.C2BA8B1E-AD7F-486D-B22E-6381108EEE0C</base:localId>
          <base:namespace>BE.NGI.ITGI-Vref</base:namespace>
        </base:Identifier>
      </bu-base:inspireId>
      <bu-core2d:geometry2D>
        <bu-base:BuildingGeometry2D xmlns:bu-base="http://inspire.ec.europa.eu/schemas/bu-base/3.0">
          <bu-base:geometry>
            <gml:Polygon gml:id="BU-CORE2D_BUILDING_0f14e287-6f2b-42b0-a55e-4e2a760e903d_BU-CORE2D_GEOMETRY2D" srsName="EPSG:900913">
              <gml:exterior>
                <gml:LinearRing>
                  <gml:posList>507054.821 6623603.419 507058.563 6623593.469 507051.137 6623590.659 507047.393 6623600.609 507054.821 6623603.419</gml:posList>
                </gml:LinearRing>
              </gml:exterior>
            </gml:Polygon>
          </bu-base:geometry>
          <bu-base:referenceGeometry>true</bu-base:referenceGeometry>
          <bu-base:horizontalGeometryReference xmlns:xlink="http://www.w3.org/1999/xlink" xlink:href="http://inspire.ec.europa.eu/codelist/HorizontalGeometryReferenceValue/roofEdge"/>
          <bu-base:verticalGeometryReference xmlns:xlink="http://www.w3.org/1999/xlink" xlink:href="http://inspire.ec.europa.eu/codelist/ElevationReferenceValue/generalRoofEdge"/>
          <bu-base:horizontalGeometryEstimatedAccuracy uom="m">1.25</bu-base:horizontalGeometryEstimatedAccuracy>
          <bu-base:verticalGeometryEstimatedAccuracy uom="m">0.35</bu-base:verticalGeometryEstimatedAccuracy>
        </bu-base:BuildingGeometry2D>
      </bu-core2d:geometry2D>
    </bu-core2d:Building>
  </wfs:member>
</wfs:FeatureCollection>
```

A quick fix would be to change the `crs:UsedGeographicCRS` definition of epsg:3857 to `google_maps_geographiccrs` in this file:

https://raw.githubusercontent.com/deegree/deegree3/master/deegree-core/deegree-core-cs/src/main/resources/org/deegree/cs/persistence/deegree/d3/config/crs-definitions.xml

Thanks,
Christophe

**Comment by jarped** [comment]
It looks to me like deegree indeed implements 3857 wrong. The following quote is from "1.3.3.2 Popular Visualisation Pseudo Mercator" in EPSG Guidance Note 7-2 (http://www.iogp.org/pubs/373-07-2.pdf)

> This method is utilised by some popular web mapping and visualisation applications. It 
applies standard Mercator (Spherical) formulas (section 1.3.3.1 above) to ellipsoidal coordinates and the sphere radius is taken to be the semi-major axis of the ellipsoid.

Also look at the gml definition at epsg-registry.org:
http://epsg-registry.org/export.htm?gml=urn:ogc:def:crs:EPSG::3857
refers to 
http://epsg-registry.org/export.htm?gml=urn:ogc:def:coordinateOperation:EPSG::3856
which refers to
http://epsg-registry.org/export.htm?gml=urn:ogc:def:method:EPSG::1024 
which refers to the above mentioned document.

**Comment by tfr42** [comment]
Can you check if PR #663 is fixing this bug? This PR will be included in the next release which will be RC2.

**Comment by f-christophe** [comment]
I've replaced both `projection-definitions.xml` and `crs-definitions.xml` files with those from PR #663 and the problem is fixed on a 3.3.17 install.

**Comment by copierrj** [comment]
merged relevant pull requests

**Comment by stephanr** [comment]
fixed with #663 and #671
